_blog-src/_posts/2016-04-08-data-encryption.adoc

```diff
@@ -0,0 +1,142 @@
+
+{empty}5. $$$ Profit!
+--
+
+
+== Writing Your Own Encryption Plugin
+
+One of the major benefits of working with an open source API management platform, and open source projects in general, is the ability to read and modify the source code to make it work for your business needs. In apiman, we make it easy for you to write your own implementation of the data encrypter by http://www.apiman.io/latest/developer-guide.html#_creating_a_plugin[creating a plugin that encrypts data].
+
+Once your basic plugin is set up, you can then begin working on the Java component of your plugin that contains the functionality.
+
+Every encryption plugin must implement the `IDataEncrypter` interface like this:
+
+```java
+public class SuperDataEncrypter implements IDataEncrypter {
+  /**
+  * Constructor.
+  */
+  public SuperDataEncrypter(Map<String, String> config) {
+    thing = config.get("thing"); //$NON-NLS-1$
+
+    if (thing == null) {
+      throw new RuntimeException("Missing configuration property: apiman-manager.config.thing"); //$NON-NLS-1$
+    }
+  }
+}
+```
+
+
+Here, we are passing a configuration object that gets created in the `ConfigFactory` we discussed above. Once the `SuperDataEncrypter` class is invoked, the object is checked for a "thing" key. If it is null, a runtime exception is thrown. Pretty, straightforward so far, right?
+
+When saving data into Elasticsearch, the data is _pre-marshalled_ and _post-marshalled_, which checks to see if you have a _pre-encryption_ and _post-encryption_ method in your class, respectively. You can pretty much guess what our next step will be:
+
+```java
+public class SuperDataEncrypter implements IDataEncrypter {
+  /**
+  * Constructor.
+  */
+  public SuperDataEncrypter(Map<String, String> config) {
+    thing = config.get("thing"); //$NON-NLS-1$
+
+    if (thing == null) {
+      throw new RuntimeException("Missing configuration property: apiman-manager.config.thing"); //$NON-NLS-1$
+    }
+  }
+
+
+  /* (non-Javadoc)
+   * @see io.apiman.common.util.crypt.IDataEncrypter#encrypt(java.lang.String)
+   */
+  @Override
+  public String encrypt(String plainText) {
+      return AesEncrypter.encrypt(secretKey, plainText);
+  }
+
+  /* (non-Javadoc)
+   * @see io.apiman.common.util.crypt.IDataEncrypter#decrypt(java.lang.String)
+   */
+  @Override
+  public String decrypt(String encryptedText) {
+      return AesEncrypter.decrypt(secretKey, encryptedText);
+  }
+}
+```
+
+From there, you can use your `SuperDataEncrypter` class to write the functionality you need. For an example of an implementation you can take a look at the `AesDataEncrypter` class we discussed earlier. You should be up and running in no time!
+
+NOTE: If you are kind enough to want to share your plugin with the rest of the apiman community, create a pull request on the main https://github.com/apiman/apiman[apiman repository]. Contributors, please don't forget to write a unit test for any plugin you write (one that actually passes, kthx!).
```

Comment:
NOTE: If you are kind enough to want to share your plugin with the rest of the apiman community, create a pull request on the https://github.com/apiman/apiman-plugins[apiman plugins repository]. Contributors, please don't forget to write a unit test for any plugin you write (one that actually passes, kthx!).